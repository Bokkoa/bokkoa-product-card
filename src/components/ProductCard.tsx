import React, { CSSProperties, createContext } from 'react'
import styles from '../styles/styles.module.css'
import { useProduct } from '../hooks/useProduct'
import { InitialValues, Product, ProductCardHandlers, ProductContextProps, onChangeArgs } from '../interfaces/interfaces'


export const ProductContext = createContext({} as ProductContextProps)
const { Provider } = ProductContext


export interface ProductCardProps {
  product: Product
  // children?: ReactElement | ReactElement[]
  children: (args: ProductCardHandlers) => JSX.Element
  className?: string
  style?: CSSProperties
  onChange?: ( args: onChangeArgs ) => void;
  value?: number;
  initialValues?: InitialValues
}


export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: ProductCardProps) => {

  const { counter, increaseBy, maxCount, reset, isMaxCountReached } = useProduct({ onChange, product, value, initialValues })

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount
      }}
    >
      <div 
        className={`${styles.productCard} ${className}`}
        style={style}
      >
        {/* STATE INITIALIZER AND FUNCTION CHILD PATTERN IMPLEMENTED */}
        {children({
          count: counter,
          maxCount: initialValues?.maxCount,
          product,
          increaseBy,
          isMaxCountReached,
          reset
        })}
      </div>
    </Provider>
  )
}
