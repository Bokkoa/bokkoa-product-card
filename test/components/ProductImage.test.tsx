import * as React from 'react';
import renderer from 'react-test-renderer'
import ProductCard, { ProductImage } from '../../src/components'
import { product2 } from '../data/products';

describe('ProductImage', () => { 

  test('should show the component with the image url correctly', () => { 
    const wrapper = renderer.create(
      <ProductImage img='https://i.insider.com/60638bd66183e1001981966a?width=1136&format=jpeg'/>
    )

    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  test('should show the component with product\'s image', () => { 
    const wrapper = renderer.create(
      <ProductCard product={product2}>
        {
          () => (
            <ProductImage />
          )
        }
      </ProductCard>
    )
    expect(wrapper.toJSON()).toMatchSnapshot()

  })
})