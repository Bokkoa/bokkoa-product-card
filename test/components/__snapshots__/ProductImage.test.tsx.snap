// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage should show the component with product's image 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product illustration"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`ProductImage should show the component with the image url correctly 1`] = `
<img
  alt="Product illustration"
  className="productImg undefined"
  src="https://i.insider.com/60638bd66183e1001981966a?width=1136&format=jpeg"
/>
`;
